### ISSUE TITLE
Timer threw unhandled excption (NullPointerException)

### ISSUE THREAD

**Issue by eugeneventer** [gocd/gocd#1826]
Hi there,

Found this in my logs, which meant some timed pipelines did not get triggerred:
```
2016-01-23 12:00:00,383 ERROR [DefaultQuartzScheduler_Worker-3] JobRunShell:211 - Job CruiseTimers.[redactedjobname] threw an unhandled Exception: 
java.lang.NullPointerException
        at com.thoughtworks.go.server.service.SchedulingCheckerService.timerTriggerCheckers(SchedulingCheckerService.java:172)
        at com.thoughtworks.go.server.service.SchedulingCheckerService.canTriggerPipelineWithTimer(SchedulingCheckerService.java:86)
        at com.thoughtworks.go.server.service.TimedBuild.canProduce(TimedBuild.java:57)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.newProduceBuildCause(BuildCauseProducerService.java:165)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.newProduceBuildCause(BuildCauseProducerService.java:161)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.timerSchedulePipeline(BuildCauseProducerService.java:144)
        at com.thoughtworks.go.server.service.TimerScheduler$SchedulePipelineQuartzJob.execute(TimerScheduler.java:162)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:525)
2016-01-23 12:00:00,384 ERROR [DefaultQuartzScheduler_Worker-3] ErrorLogger:2185 - Job (CruiseTimers.[redactedjobname] threw an exception.
org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: java.lang.NullPointerException]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:525)
Caused by: java.lang.NullPointerException
        at com.thoughtworks.go.server.service.SchedulingCheckerService.timerTriggerCheckers(SchedulingCheckerService.java:172)
        at com.thoughtworks.go.server.service.SchedulingCheckerService.canTriggerPipelineWithTimer(SchedulingCheckerService.java:86)
        at com.thoughtworks.go.server.service.TimedBuild.canProduce(TimedBuild.java:57)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.newProduceBuildCause(BuildCauseProducerService.java:165)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.newProduceBuildCause(BuildCauseProducerService.java:161)
        at com.thoughtworks.go.server.scheduling.BuildCauseProducerService.timerSchedulePipeline(BuildCauseProducerService.java:144)
        at com.thoughtworks.go.server.service.TimerScheduler$SchedulePipelineQuartzJob.execute(TimerScheduler.java:162)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more
```
Could you take a look and let me know how I can rectify this?

Thanks,
Eugene.

**Comment by xli** [comment]
@eugeneventer do you see this error all the time? 
It looks like there is problem when Go server tries to trigger a build by timer, but it should get recovered in next run.
If you see a lot of this errors in log, you can try restart Go server.

**Comment by arvindsv** [comment]
Could it be that the provided timer configuration is somehow wrong?

**Comment by eugeneventer** [comment]
@xli and @arvindsv I see this error occasionally - sometimes it seems to work. This is probably allright for pipelines that are scheduled to run e.g hourly, but I have the same issue with a pipeline that only runs once a week - the timer config is: 0 0 20 ? * FRI
The pipeline threw the exception, which means the pipeline did not even run on Friday...

**Comment by arvindsv** [comment]
@xli and @jyotisingh: What do you think is happening? Can the pipeline config be null? Or getFirstStageConfig? I don't see how that can happen. These are in-memory objects.

If it failed on the second of these lines:

```
String pipelineName = CaseInsensitiveString.str(pipelineConfig.name());
String stageName = CaseInsensitiveString.str(pipelineConfig.getFirstStageConfig().name());
```
then, pipelineConfig cannot be null. It'll have to be `getFirstStageConfig` which is null. Which still doesn't make sense.

**Comment by eugeneventer** [comment]
@arvindsv just a thought, but could changing the config at the same time cause this?

I note 'reload timer' entries in the log when the config is updated. Could it be that the timer is reloaded when it is triggered or something?

Just a stab in the dark to try and help out ;)

**Comment by xli** [comment]
I couldn't figure out what's wrong. @eugeneventer could you parse your pipeline configuration (exclude any sensitive info) here? So that I can try reproduce this problem.

**Comment by arvindsv** [comment]
@eugeneventer: I'm not sure. Theoretically, it shouldn't get affected by it, because a change in config is supposed to throw away the old timers and create new ones. But, I suspect you're right. It has something to do with a change. But, it's just a hunch. I could be wrong. As @xli said, maybe more information might help. Which version of the Go Server is this on?

**Comment by eugeneventer** [comment]
@arvindsv I'm running on the latest version - 16.1

**Comment by eugeneventer** [comment]
Another thing to note is that we have about 200 pipelines triggering hourly (at the same time).

**Comment by jyotisingh** [comment]
The statement throwing the exception is 
```
String stageName = CaseInsensitiveString.str(pipelineConfig.getFirstStageConfig().name());
```
so that makes the stage to be null.
@eugeneventer - Can you tell if the pipelines which are affected by this error are based on a template? If thats not the case, It would probably help if you could share the pipeline-config as @xli suggested.

**Comment by eugeneventer** [comment]
@jyotisingh yes, these pipelines are indeed based on a template - the same template is currently used by all of these timed pipelines.

See the pipeline config below (with redacted bits):
```
    <pipeline name="some-pipe-name" template="the-template">
      <params>
        <param name="GO_PARAM1">string1</param>
        <param name="GO_PARAM2">string2</param>
        <param name="GO_PARAM3">string3</param>
        <param name="GO_PARAM4">1</param>
      </params>
      <timer onlyOnChanges="true">0 0 * * * ?</timer>
      <materials>
        <git url="gitmirror" branch="somebranch" />
      </materials>
    </pipeline>
```

**Comment by xli** [comment]
@eugeneventer no stage is configured? or they are not included.

**Comment by eugeneventer** [comment]
@xli the stages are configured in the template.

**Comment by eugeneventer** [comment]
@jyotisingh and @xli here's the template pipeline's config:
```
    <pipeline name="the-template">
      <stage name="stage-name" fetchMaterials="false">
        <approval type="manual" />
        <jobs>
          <job name="jobname">
            <tasks>
              <exec command="python">
                <arg>/path/to/some/script.py</arg>
                <arg>-s</arg>
                <arg>#{GO_PARAM1}</arg>
                <arg>-e</arg>
                <arg>#{GO_PARAM2}</arg>
                <arg>-t</arg>
                <arg>#{GO_PARAM3}</arg>
              </exec>
            </tasks>
          </job>
        </jobs>
      </stage>
    </pipeline>
```

**Comment by eugeneventer** [comment]
Hi guys,

Any progress here?

Thanks!

**Comment by xli** [comment]
@eugeneventer sorry for late. 
I still can not reproduce problem. I've tried 1000 pipelines triggered every minutes. 

The only possible I guess is pipeline configuration changed when Go tried to trigger pipeline.
However, I could not reproduce it with every minutes triggered pipelines.
So I can't imaging what you do can trigger this problem.
Could you provide more information? Like logs when the problem happened. Or a list of things maybe done when the pipelines are triggered. Or do you have a script periodically update Go configuration.

**Comment by jyotisingh** [comment]
Was able to reproduce the issue with the template and pipeline provided by @eugeneventer. The issue occurs only when such a pipeline is edited via the pipeline config-api and the next timer trigger fails with the error posted in the original issue. 
Submitted a pr (#1921) for this.

**Comment by eugeneventer** [comment]
@jyotisingh yes! I'm periodically updating the pipelines via the API, using a script! I can confirm that this was the issue, as I've disabled the update script and the exceptions have stopped ;)

Thank you very much and good job on tracking down the issue! :)

**Comment by jyotisingh** [comment]
@eugeneventer - the fix for this has been merged and would be available as a part of 16.2 which is due next week(monday?). You would be able to enable the update script after an upgrade. Thanks for the reporting the bug.

**Comment by rajiesh** [comment]
verified on 16.2.0-3024(0946fc7a9a54555f437fda398fafde0fbc7985bd), works as expected
